Update scripts to use the GW-RT global namelist

dpsarmie opened the pull request:
# Description
------
### The following is a Google Doc with new variables that were added to the nml template with this PR: [Google Sheets link](https://docs.google.com/spreadsheets/d/1xhzXDL__4OJCOz1bPzC1OiooUBRtdz8y-CPzX5uhpv8/edit?usp=sharing)
------
This PR changes the ways the input.nml file is created. Instead of creating a namelist, the variables that are set locally will now use the UFS `global_control.nml.IN` template and the atparse function to generate the namelist.

This should allow the GW branch to more easily test changes in the UFS regression test framework. It will also allow GW to more easily introduce UFS changes into GW.
 
The functionality in `parsing_namelists_FV3.sh` is mostly unchanged. Since variables are still saved as `local`, the parsing of the template will occur in this script. Sections in the namelist that were previously not generated (stoch physics options for example) will be set and parsed since the regression test template has these placeholders active. As options are turned off and on, ! will be used in the namelist to comment out sections that are not used.

The MOM6 variables have been moved here since the template place holders are active in global_control.nml.IN.

Since the stoch physics options are active, the SKEB definitions were also modified in the config.efcs files. If the first value in SKEB is not -999., then the stoch physics subroutine will turn SKEB on (do_skeb set to 1).

The FHZERO variable was also renamed to reflect what was in the namelist template.

Previous PR #2733 was closed since my old branch was stale and a new branch was created.

### Unresolved issues and input needed from GW group before final code review
There are some variables that were not set in the original scripts that have been added. These were set to the UFS defaults. 
The tests I've conducted were to generate the namelists but more tests should be done to make sure that outputs have not changed.


Resolves #2731 
Resolves https://github.com/ufs-community/ufs-weather-model/issues/1664

Outstanding dependencies: 
https://github.com/ufs-community/ufs-weather-model/pull/2425


# Type of change
- [ ] Bug fix (fixes something broken)
- [X] New feature (adds functionality)
- [ ] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? YES
- Does this change require a documentation update? YES
- Does this change require an update to any of the following submodules? YES
  - [ ] EMC verif-global <!-- NOAA-EMC/EMC_verif-global#1234 -->
  - [ ] GDAS <!-- NOAA-EMC/GDASApp#1234 -->
  - [ ] GFS-utils <!-- NOAA-EMC/gfs-utils#1234 -->
  - [ ] GSI <!-- NOAA-EMC/GSI#1234 -->
  - [ ] GSI-monitor <!-- NOAA-EMC/GSI-Monitor#1234 -->
  - [ ] GSI-utils <!-- NOAA-EMC/GSI-Utils#1234 -->
  - [ ] UFS-utils <!-- ufs-community/UFS_UTILS#1234 -->
  - [X] UFS-weather-model  ufs-community/ufs-weather-model/pull/2425
  - [ ] wxflow <!-- NOAA-EMC/wxflow#1234 -->

# How has this been tested?
Tested namelist generation with C48ATM and C48S2S configurations. Further testing will need to be done to ensure intended funcitonality. 

# Checklist
- [ ] Any dependent changes have been merged and published
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have documented my code, including function, input, and output descriptions
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] This change is covered by an existing CI test or a new one has been added
- [ ] I have made corresponding changes to the system documentation if necessary

WalterKolczynski-NOAA commented:
Not a fan of the hiding, at least not how it is implemented here. All the variables are still set by the parsing script, it makes the template more difficult to read, and doesn't even remove it from the rendered namelist, just comments the lines out. Given that, I'd rather just leave all the settings in the namelist with the values that turn them off.

dpsarmie commented:
Hey Walter, 
Agreed, this wasn't my first choice but I didn't want to completely overhaul the scripts on the first go. 
We could think about also integrating the default_vars.sh script into this workflow, but I didn't know how you all would feel about saving the environmental variables globally vs locally. If it is allowed, then you could in theory run the default_vars.sh script to set any variables that aren't defined by the user's previous scripts before populating the namelist. That should cut down on the amount of bulk in the parsing_fv3 script.

Splitting up the sections into different scripts could also work, but you run into the issue of not being able to easily pass variables between scripts if they are saved as locals (correct me if I'm wrong here).

Also open to your suggestions on how to best make this fit within the GW system.

dpsarmie commented:
Here is a doc with the differences between variable values for GW and RT for a C48 atm only case: [GW-RT differences](https://docs.google.com/spreadsheets/d/1PFgYllDaMoNw9tfkYlTSHDxdGXmNgu5miVCp-IHD3fo/edit?usp=sharing)

We're leaning towards just changing the RT baselines to match what GW has, but we can discuss this if needed.

dpsarmie commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -80,690 +80,531 @@
+local NPZP=${LEVS} #levp
+local GFS_DWINDS=${gfs_dwinds}
+
+local FHZERO=${FHZER}
```
Old script saved this as FHZER (ush/parsing_namelists_FV3.sh L196).  Kept the convention for now.

WalterKolczynski-NOAA replied:
We should just update the other instances:
```
parm/config/gfs/config.efcs:export FHZER=6
parm/config/gefs/config.fcst:export FHZER=@FHZER@
parm/config/gefs/yaml/defaults.yaml:  FHZER: 6
ush/parsing_namelists_FV3_nest.sh:  fhzero       = ${FHZER}
```

dpsarmie committed:
Finalized parsing changes

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
adjust MOM6 nml changes

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

WalterKolczynski-NOAA reviewed (changes requested):
Structurally it looks fine. Didn't verify that the default for every individual setting is unchanged. Just take care of the `FHZER` thing and I'm happy with this.

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Fixes FHZER calls/definitions

WalterKolczynski-NOAA commented:
@dpsarmie Merge in develop again and I think this is ready for CI

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie commented:
I was still running tests on my end as a final check but should be done before lunch. We should still talk about the final pathing to the template (just added a review for visibility).

dpsarmie committed:
Fixes typo in PERT_MP

Added a missing period

dpsarmie commented:
@WalterKolczynski-NOAA Ok, good to go on my end pending the needed pathing change.  

I added a google sheet to the PR desc to document the variables that are new to the GW template and all pre-existing values are unchanged.

It'll be easier see the changes in case they are needed.

dpsarmie committed:
Revert conflicts

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Resolve conflict

WalterKolczynski-NOAA commented:
> I was still running tests on my end as a final check but should be done before lunch. We should still talk about the final pathing to the template (just added a review for visibility).

The templates should be linked into `parm/ufs/fv3` by `sorc/link_workflow.sh`.

dpsarmie commented:
> The templates should be linked into `parm/ufs/fv3` by `sorc/link_workflow.sh`.

Ok, got it going but ran into another issue. The hash GW is using ([63ace62](https://github.com/ufs-community/ufs-weather-model/commit/63ace62)) is using a namelist template with the [HIDE_X] tags and is incompatible with the script based off the current namelist template.

I can add the HIDE_X definitions back in and then they would be removed next time GW updates the UFS hash? We could also just sit on this PR until the next hash update.

Let me know what you think makes sense.

WalterKolczynski-NOAA commented:
#3190 updates the ufs model hash, so it shouldn't be too long a wait.

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Add links to nml

dpsarmie commented:
I'll start taking a look at this for a fix

dpsarmie committed:
Fix single quotes and PERT_CLDS setting

Single quotes have been forced on all filenames in the namelist.
PERT_CLDS has also been set to .false. unless the DO_SPPT flag is
on, then PRET_CLDS is set to .true.

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Modify link_workflow to try to address CI failures

Changes were made to sorc/link_workflow.sh and
ush/parsing_namelists_FV3.sh to try and address the CI
failures. There were still some discrepancies in the
namelists where XR_CNVCLD was being used but was not
declared in the template and INCREMENT_FILE_ON_NATIVE_GRID
was in the template but the default value might have been
incompatible with some of the CI cases.

If the CI tests pass, then a more permanent solution should be
found and the code will need to be cleaned up before the
final merge into the GW repo.

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Modify defaults for GEFS CI

The default values of the following variables have been
changed:
  XR_CNVCLD has been reverted to the GW default.
  INCREMENT_FILE_ON_NATIVE_GRID has been changed to .false.
  all LAND_IAU variables have been added
  SHUM default has been set to -999 because it was turning on even when
     do_shum was set to .false.

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Change remaining FHZER to FHZERO

dpsarmie committed:
Fix C96mx100_sfs CI case

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Fix SFS SKEB variable

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie commented on ush/forecast_predet.sh:
```diff
@@ -146,7 +146,7 @@
   ${NLN} "${DATArestart}/FV3_RESTART" "${DATA}/RESTART"
 
-  FHZER=${FHZER:-6}
+  FHZERO={FHZERO:-6}
```
This has been fixed

DavidHuber-NOAA reviewed (dismissed):
Changes look good.  Thanks for this work @dpsarmie.  Approve pending the completion of a few tests.  Before starting these tests, please merge in develop.

Next, could you please run a C48_S2SW test case on this branch then run the same case with develop?  I would like to compare products from the two runs to verify everything is 1-to-1.

Also, when running these test cases, please set `export KEEPDATA=YES` in config.base so that the run directories are maintained for both this branch and develop.  Once complete, please point me to your COMROOT directory so I can run diffs on the resulting run directories.

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie commented:
Hey @DavidHuber-NOAA,

My COMROOT on Hera is here: `/scratch1/NCEPDEV/stmp2/Daniel.Sarmiento/RUNTESTS/COMROOT`

There's other CI test output in there that were created when I last pushed changes a few days ago (in case you want to look), but the `c48s2sw` case is using the latest version of the branch with the most up-to-date GW changes.

Let me know if you need anything else.

aerorahul commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -80,673 +80,501 @@ else
-  oz_phys      = .false.
-  oz_phys_2015 = .true.
-EOF
+    local OZ_PHYS_NEW=.false. #oz_phys
```
Just noting here that these suites are no longer supported in the global-workflow, so may want to just remove these cases.

dpsarmie replied:
Ok, I can pull out the variable definitions and delete the case block.

dpsarmie replied:
Changes have been made and tested with the C48ATM CI test.

dpsarmie committed:
Remove outdated CCPP_SUITE case block

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

DavidHuber-NOAA commented:
@dpsarmie @aerorahul apologies for this taking as long as it did.  I have finished comparing your atmosphere, ice, ocean, and wave products against develop and they are identical.  I did not a few differences in the input namelists and I think most of them are fine.  And in fact your PR also produces a `land_iau_nml` (in which `do_land_iau=.false., as I believe it should be), which may resolve the bug that we were seeing in https://github.com/NOAA-EMC/global-workflow/issues/3357 (FYI @DavidGrumm-NOAA).

However, there is one change to the namelists that I am not sure is correct.  For the `c48s2sw` case, I see in
`/scratch1/NCEPDEV/stmp2/Daniel.Sarmiento/RUNTESTS/COMROOT/c48s2sw/gfs.20210323/12/conf/ufs.input.nml`
under `gfs_physics_nml` that a few `iau` variables are set, namely
```
  iaufhrs      = 6,
  iau_delthrs  = 6 
  iau_inc_files= ''
  iau_drymassfixer = .false.
```

I don't think there should be any IAU variables set for a forecast-only case.  Can you confirm that @aerorahul?  It may not matter since the `iau_inc_files` variable is empty, which may implicitly disable IAU in the forecast (not sure about this).

I think I will repeat this case for `c48mx500_hyb`.  If the results are identical for that case, then I will be satisfied with 1-to-1 results.  I will compare against the products in `/scratch1/NCEPDEV/stmp2/Daniel.Sarmiento/RUNTESTS/COMROOT/c48mx500_hyb`.

DavidHuber-NOAA commented:
@dpsarmie I have finished running and comparing the `C48mx500_hybAOWCDA` case with your new namelists against develop and the results are identical.  However, there are a few differences I noted in the resulting namelists.  Similar to the `C48_S2SW` case, there were some variables defined in your namelists that were not defined in develop.  Below is the output of `test/f90nmlcmp.sh /scratch1/NCEPDEV/global/David.Huber/para_nml/COMROOT/C48mx500_hybAOWCDA_baseline/gdas.20210325/00/conf/ufs.input.nml /scratch1/NCEPDEV/global/David.Huber/para_nml/COMROOT/C48mx500_hybAOWCDA_new_nml/gdas.20210325/00/conf/ufs.input.nml`:
```
comparing: C48mx500_hybAOWCDA_new_nml/gdas.20210325/00/conf/ufs.input.nml | C48mx500_hybAOWCDA_baseline/gdas.20210325/00/conf/ufs.input.nml
------------------------------------------------------------------------------------------------------------------
fv_core_nml:
  increment_file_on_native_grid : False | UNDEFINED
gfs_physics_nml:
    fscav_aero : ['*:0.3', 'so2:0.0', 'msa:0.0', 'dms:0.0', 'nh3:0.4', 'nh4:0.6', 'bc1:0.6', 'bc2:0.6', 'oc1:0.4', 'oc2:0.4', 'dust1:0.6', 'dust2:0.6', 'dust3:0.6', 'dust4:0.6', 'dust5:0.6', 'seas1:0.5', 'seas2:0.5', 'seas3:0.5', 'seas4:0.5', 'seas5:0.5'] | UNDEFINED
  dogp_sgs_cnv : True | UNDEFINED
  use_med_flux : False | UNDEFINED
        cpllnd : False | UNDEFINED
    cpllnd2atm : False | UNDEFINED
     lndp_type : 0 | UNDEFINED
    n_var_lndp : 0 | UNDEFINED
        lcnorm : False | UNDEFINED
       pert_mp : False | UNDEFINED
  pert_radtend : False | UNDEFINED
     pert_clds : False | UNDEFINED

nam_stochy:
            skeb : 0 | UNDEFINED
      iseed_skeb : 0 | UNDEFINED
        skeb_tau : -999.0 | UNDEFINED
     skeb_lscale : -999.0 | UNDEFINED
        skebnorm : 1 | UNDEFINED
      skeb_npass : 30 | UNDEFINED
       skeb_vdof : 5 | UNDEFINED
            shum : -999.0 | UNDEFINED
      iseed_shum : 0 | UNDEFINED
        shum_tau : -999.0 | UNDEFINED
     shum_lscale : -999.0 | UNDEFINED
            sppt : -999.0 | UNDEFINED
      iseed_sppt : 0 | UNDEFINED
        sppt_tau : -999.0 | UNDEFINED
     sppt_lscale : -999.0 | UNDEFINED
      sppt_logit : True | UNDEFINED
   sppt_sfclimit : True | UNDEFINED
    use_zmtnblck : True | UNDEFINED
       pbl_taper : [0, 0, 0, 0.125, 0.25, 0.5, 0.75] | UNDEFINED
         ocnsppt : [0.8, 0.4, 0.2, 0.08, 0.04] | UNDEFINED
  ocnsppt_lscale : [500000.0, 1000000.0, 2000000.0, 2000000.0, 2000000.0] | UNDEFINED
     ocnsppt_tau : [21600.0, 259200.0, 2592000.0, 7776000.0, 31536000.0] | UNDEFINED
   iseed_ocnsppt : 0 | UNDEFINED
            epbl : [0.8, 0.4, 0.2, 0.08, 0.04] | UNDEFINED
     epbl_lscale : [500000.0, 1000000.0, 2000000.0, 2000000.0, 2000000.0] | UNDEFINED
        epbl_tau : [21600.0, 259200.0, 2592000.0, 7776000.0, 31536000.0] | UNDEFINED
      iseed_epbl : 0 | UNDEFINED
nam_sfcperts:
        lndp_type : 0 | UNDEFINED
  lndp_model_type : 0 | UNDEFINED
         lndp_tau : 21600 | UNDEFINED
      lndp_lscale : 500000 | UNDEFINED
       iseed_lndp : 0 | UNDEFINED
    lndp_var_list : XXX | UNDEFINED
    lndp_prt_list : -999 | UNDEFINED
land_iau_nml:
  land_iau_nml : {'do_land_iau': False, 'land_iau_fhrs': [3, 6, 9], 'land_iau_delthrs': 6, 'land_iau_inc_files': ['sfc_inc', ''], 'lsoil_incr': 3, 'land_iau_filter_increments': False, 'land_iau_upd_stc': True, 'land_iau_upd_slc': True, 'land_iau_do_stcsmc_adjustment': True, 'land_iau_min_t_increment': 0.0001} | UNDEFINED
```
I think that the majority of these are defaults.  Can you confirm if any of these are **_not_** the defaults?  In particular, ` pert_radtend : False`.  Comparing the log forecast log files, I see in develop that the UFS seems to initialize this to `True`:
```
 0:   pert_radtend    :  T
```

Also, I noticed a few duplicate entries in the new namelist.   Namely, `iau_inc_files` and `iau_drymass_increments`, both under `gfs_physics_nml`.  These should probably only be printed once.

---------------------------
And for the C48_S2SW case:
```
comparing: /scratch1/NCEPDEV/stmp2/Daniel.Sarmiento/RUNTESTS/COMROOT/c48s2sw/gfs.20210323/12/conf/ufs.input.nml | gfs.20210323/12/conf/ufs.input.nml
----------------------------------------------------------------------------------------------------------------------------------------------------
fv_core_nml:
  increment_file_on_native_grid : False | UNDEFINED
gfs_physics_nml:
             fscav_aero : ['*:0.3', 'so2:0.0', 'msa:0.0', 'dms:0.0', 'nh3:0.4', 'nh4:0.6', 'bc1:0.6', 'bc2:0.6', 'oc1:0.4', 'oc2:0.4', 'dust1:0.6', 'dust2:0.6', 'dust3:0.6', 'dust4:0.6', 'dust5:0.6', 'seas1:0.5', 'seas2:0.5', 'seas3:0.5', 'seas4:0.5', 'seas5:0.5'] | UNDEFINED
           dogp_sgs_cnv : True | UNDEFINED
                iaufhrs : 6 | UNDEFINED
            iau_delthrs : 6 | UNDEFINED
          iau_inc_files :  | UNDEFINED
       iau_drymassfixer : False | UNDEFINED
           use_med_flux : False | UNDEFINED
                 cpllnd : False | UNDEFINED
             cpllnd2atm : False | UNDEFINED
              lndp_type : 0 | UNDEFINED
             n_var_lndp : 0 | UNDEFINED
                 lcnorm : False | UNDEFINED
                pert_mp : False | UNDEFINED
           pert_radtend : False | UNDEFINED
              pert_clds : False | UNDEFINED
  iau_filter_increments : False | UNDEFINED
nam_stochy:
            skeb : 0 | UNDEFINED
      iseed_skeb : 0 | UNDEFINED
        skeb_tau : -999.0 | UNDEFINED
     skeb_lscale : -999.0 | UNDEFINED
        skebnorm : 1 | UNDEFINED
      skeb_npass : 30 | UNDEFINED
       skeb_vdof : 5 | UNDEFINED
            shum : -999.0 | UNDEFINED
      iseed_shum : 0 | UNDEFINED
        shum_tau : -999.0 | UNDEFINED
     shum_lscale : -999.0 | UNDEFINED
            sppt : -999.0 | UNDEFINED
      iseed_sppt : 0 | UNDEFINED
        sppt_tau : -999.0 | UNDEFINED
     sppt_lscale : -999.0 | UNDEFINED
      sppt_logit : True | UNDEFINED
   sppt_sfclimit : True | UNDEFINED
    use_zmtnblck : True | UNDEFINED
       pbl_taper : [0, 0, 0, 0.125, 0.25, 0.5, 0.75] | UNDEFINED
         ocnsppt : [0.8, 0.4, 0.2, 0.08, 0.04] | UNDEFINED
  ocnsppt_lscale : [500000.0, 1000000.0, 2000000.0, 2000000.0, 2000000.0] | UNDEFINED
     ocnsppt_tau : [21600.0, 259200.0, 2592000.0, 7776000.0, 31536000.0] | UNDEFINED
   iseed_ocnsppt : 0 | UNDEFINED
            epbl : [0.8, 0.4, 0.2, 0.08, 0.04] | UNDEFINED
     epbl_lscale : [500000.0, 1000000.0, 2000000.0, 2000000.0, 2000000.0] | UNDEFINED
        epbl_tau : [21600.0, 259200.0, 2592000.0, 7776000.0, 31536000.0] | UNDEFINED
      iseed_epbl : 0 | UNDEFINED
nam_sfcperts:
        lndp_type : 0 | UNDEFINED
  lndp_model_type : 0 | UNDEFINED
         lndp_tau : 21600 | UNDEFINED
      lndp_lscale : 500000 | UNDEFINED
       iseed_lndp : 0 | UNDEFINED
    lndp_var_list : XXX | UNDEFINED
    lndp_prt_list : -999 | UNDEFINED
land_iau_nml:
  land_iau_nml : {'do_land_iau': False, 'land_iau_fhrs': [3, 6, 9], 'land_iau_delthrs': 6, 'land_iau_inc_files': ['sfc_inc', ''], 'lsoil_incr': 3, 'land_iau_filter_increments': False, 'land_iau_upd_stc': True, 'land_iau_upd_slc': True, 'land_iau_do_stcsmc_adjustment': True, 'land_iau_min_t_increment': 0.0001} | UNDEFINED
```

Similar to above, are any of these entries not the defaults?  In particular, I don't think IAU variables should be set for this experiment, i.e. `iaufhrs : 6`, `iau_delthrs : 6`.  This results in a message from the UFS that differs from develop:
new namelist
```
 0:   model init,iaufhrs=   6.00000000000000       -1.00000000000000
 0:   -1.00000000000000       -1.00000000000000       -1.00000000000000
 0:   -1.00000000000000       -1.00000000000000
```

develop
```
  0:   model init,iaufhrs=  -1.00000000000000       -1.00000000000000
  0:   -1.00000000000000       -1.00000000000000       -1.00000000000000
  0:   -1.00000000000000       -1.00000000000000
```

Lastly, I noticed that a couple of entries are duplicated in the new namelists.  These were the same as the `C48mx500_hybAOWCDA` case (`iau_inc_files` and `iau_drymass_increments`).

dpsarmie commented:
@DavidHuber-NOAA thanks for going through this. 
I will go ahead and remove any duplicates.
____________________________________________________
**For the question about default values:**
I first cross-referenced what was already in the `pasring_namelist_FV3.sh` script. For example, `pert_radtend` is set to **False** only when SPPT is being used.  When `pert_radtend` was not being set (any non-SPPT run), I used the values present in the UFS `default_vars.sh` script (which also sets this value to false for all configurations as default).

**_I did not go into the submodules themselves to fetch the default values that are being set if the variable is not defined._** I think this is where the disconnect it occurring.

At the end of the day, this is for GW and separate from the UFS default_vars.sh and regression test workflow. So I can go back and set these up to match the defaults that were being set when variables are undefined. (For this example, have pert_radtend set to True unless DO_SPPT is YES.) 
__________________________________________________
**For the IAU variables:**
Same reasoning for default values apply here. We can set the values to whatever you all feel is correct for GW (set to -1 instead of 6).
Also, there is the possibility of adding a `@[HIDE_IAU]` tag for these variables. This can be switched to `!` when IAU is not required. The variables will then be read in as undefined by UFS if there's a worry that there could be potential interference by having these variables defined.

I'm not 100% sure on the IAU interactions, but I can dig in further to see if this is being skipped when the IAU files are not defined.  I'm assuming that that is true but I haven't verified it.
___________________________________________________
Either way, let me know what you all prefer. I know this is a big change to the workflow so we should make sure everything is as you want it before merging. Thanks again for taking the time to work on this.

DavidHuber-NOAA commented:
I think I like the hidden IAU variables option better, but I will note that the `land_iau_nml` should still be created, just with the minimum number of variables and `do_land_iau=.false.` to prevent the EoF warning from the UFS.  Could you try that with a C48_S2SW test case?

dpsarmie committed:
Resolve conflicts P1 of 2

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Resolve conflicts P2 of 2

dpsarmie commented:
@DavidHuber-NOAA 
Here's the new batch of runs: /scratch1/NCEPDEV/stmp2/Daniel.Sarmiento/RUNTESTS/COMROOT

This uses the a template with the IAU variables commented out if IAU is not being used. I added a new HIDE_LIAU variable for the land iau trigger. (There's a copy of global_control.nml.IN in the COMROOT directory if you're interested.)

The iaufhrs in the logs looks like it matches develop and the warning for the Land IAU section doesn't seem to be there.

The duplicate variables are a byproduct of the global_control.nml.IN template. In GW's version of the UFS hash these duplicates were present but they were removed in [ee3d52](https://github.com/ufs-community/ufs-weather-model/commit/ee3d52d3620565dbf7bc4f47bd039fe6423fa5c1) so that issue will be resolved once GW catches up to that hash.

--------------------------------------
I'll have to make a PR on the UFS side to include the hide tags once you take a look at the run and sign off on it. GW will need to catch up to that hash before this PR can be merged.

Also, were you ok with using the defaults from the regression test workflow? Or should I go back and change the discrepancies (such as `pert_radtend` that you mentioned previously).

DavidHuber-NOAA commented:
@dpsarmie @aerorahul The runs look great!  Thank you.  As for the different `perf_radtend` and any similar variables, I will defer to Rahul.  This particular variable is not currently defined in develop when constructing ufs.input.nml.  The default appears to be `.true.`, but here is being set to `.false.`.  I'm not sure which we should be using.  It does not have an effect on 1-to-1 reproducibility between Daniel's branch and develop.

DavidHuber-NOAA commented:
I'm going to add the 'blocked' label while waiting for the upstream changes to the UFS.

dpsarmie commented:
Ok great! I'll get a PR going on the UFS side with the changes to the namelist template and push to get that merged asap.  Any further changes can be done on the GW side so we can iterate while GW catches up on the UFS hashes.

DavidHuber-NOAA commented:
@dpsarmie Are there any updates on the upstream UFS PR?

dpsarmie commented:
> @dpsarmie Are there any updates on the upstream UFS PR?

No updates. It should be slated to get merged next week. I'll bring this up with EPIC to try to make sure it does.

dpsarmie commented:
Hi @DavidHuber-NOAA,

The namelist was merged as part of https://github.com/ufs-community/ufs-weather-model/commit/4f0111b26d358b26722e326a7dd663499aa8b3f9.

The global workflow update to the 20250328 hash https://github.com/NOAA-EMC/global-workflow/commit/32be0cbc6b844b094d1ff89bfcb327dc97d06ff0 should have fixed the other issues (global_control.nml_IN has the Land IAU section and removes the duplicate namelist options).

I'll update this PR this week and I'll keep an eye on it so this can be merged once GW updates the UFS hash to include the 20250421 merge.

DavidHuber-NOAA commented:
Sounds good @dpsarmie, thanks!  When https://github.com/NOAA-EMC/global-workflow/pull/3587 is merged (hopefully today) that will update the hash to include UFS@4f011b.  I'll go ahead and unblock this PR now.  We can then start running RTs once conflicts are resolved.

dpsarmie committed:
Resolve conflicts 1 of 2

dpsarmie committed:
Resolve conflicts 2 of 2

dpsarmie committed:
Revert efcs / correct fcst files

dpsarmie committed:
Merge branch 'develop' into feature/GW-RT_GlobalNML

dpsarmie committed:
Add FHZERO changes back

dpsarmie committed:
Merge branch 'develop' into feature/GW-RT_GlobalNML

dpsarmie commented:
Updated to the latest commit but I will look this over one more time to make sure the new namelist options are reflected in the parsing script. I'm assuming the CIs will also be rerun for the PR before the final merge. There were many updates since this was last checked so I guess one last thorough check wouldn't hurt.

DavidHuber-NOAA commented:
Sounds good and thanks @dpsarmie.  When you are satisfied, let me know and I'll give it another review before launching CI.

dpsarmie committed:
Final parsing_namelist_FV3.sh changes

dpsarmie commented:
@DavidHuber-NOAA Ok, I think this should be good to go for a review and final testing.

I had to add a DO_LANDIAU variable to have an additional option from the current DOIAU. I figure that GW can modify that when the time comes to add in the Land IAU component to your runs.

DavidHuber-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -80,674 +80,419 @@ else
+local READ_INCREMENT=${read_increment}
+local RES_LATLON_DYNAMICS=${res_latlon_dynamics}
+local ATM_IGNORE_RST_CKSUM=.false.
+local INCREMENT_FILE_ON_NATIVE_GRID=.false.
```
I think `INCREMENT_FILE_ON_NATIVE_GRID` needs to first check if `increment_file_on_native_grid` is set.  This is set `.true.` when running JEDI DA.

```suggestion
local INCREMENT_FILE_ON_NATIVE_GRID=${increment_file_on_native_grid:-.false.}
```

DavidHuber-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -80,674 +80,419 @@ else
+local DSPHEAT=${dspheat:-".true."}
+local HYBEDMF=${hybedmf:-".false."}
+local SATMEDMF=${satmedmf-".true."}
+local ISATMEDMF=${isatmedmf-"1"}
```
I think this was a typo in the develop version:
```suggestion
local ISATMEDMF=${isatmedmf:-"1"}
```

DavidHuber-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -80,674 +80,419 @@ else
+local HYBEDMF=${hybedmf:-".false."}
+local SATMEDMF=${satmedmf-".true."}
+local ISATMEDMF=${isatmedmf-"1"}
+local LHEATSTRG=${lheatstrg-".false."}
```
Same
```suggestion
local LHEATSTRG=${lheatstrg:-".false."}
```

DavidHuber-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -80,674 +80,419 @@ else
+local REDRAG=${redrag:-".true."}
+local DSPHEAT=${dspheat:-".true."}
+local HYBEDMF=${hybedmf:-".false."}
+local SATMEDMF=${satmedmf-".true."}
```
Last one, I think:
```suggestion
local SATMEDMF=${satmedmf:-".true."}
```

dpsarmie replied:
Ok, typo fixes and adding the `increment` variable are now in the PR.

dpsarmie committed:
Add increment_file and fix typos

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

DavidHuber-NOAA reviewed (dismissed):
Looks great!

DavidHuber-NOAA commented:
Everything looks good to me.  Launching CI on Gaea and WCOSS2.

dpsarmie committed:
Rework land IAU check

DavidHuber-NOAA commented:
A number of tests are failing on WCOSS2.  @dpsarmie have talked offline about the issues.  It mostly seems to be incompatible namelist options during GFS C48 ensemble forecasts and C96 S2S forecasts.  Marking as failed for now and will relaunch when ready.

DavidHuber-NOAA committed:
Merge branch 'develop' into feature/GW-RT_GlobalNML

DavidHuber-NOAA committed:
Delete whitespace

DavidHuber-NOAA committed:
Set C48 deltim to 900 to align with skebint requirement

DavidHuber-NOAA committed:
Set deltim to 900 and increase layout for C48

dpsarmie committed:
Merge pull request #1 from DavidHuber-NOAA/fix/c48_skebint

Decrease DELTIM to 900 for C48 to accommodate SKEBINT requirement

dpsarmie committed:
Merge branch 'NOAA-EMC:develop' into feature/GW-RT_GlobalNML

DavidHuber-NOAA commented:
Offline tests of C96C48_hybatmaerosnowDA were successful on WCOSS2.  Marking ready for CI testing on Hercules.

DavidHuber-NOAA committed:
Merge branch 'develop' into feature/GW-RT_GlobalNML

DavidHuber-NOAA reviewed (approved):
Changes look good.  Just noting that I own the changes to `parm/config/*/config.ufs`.

KateFriedman-NOAA commented:
Giving CI status on WCOSS2 before signing off for the weekend: testing on WCOSS2 is nearing completion with all tests completed successfully minus the extended test which is still running (with no issues so far).
```
Fri May  2 18:52:39 UTC 2025
******** C48_ATM_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    May 02 2025 15:01:37    May 02 2025 15:56:35

******** C48mx500_3DVarAOWCDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103241800        Done    May 02 2025 15:01:39    May 02 2025 15:16:31
202103250000        Done    May 02 2025 15:01:39    May 02 2025 16:36:22

******** C48mx500_hybAOWCDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103241800        Done    May 02 2025 15:01:41    May 02 2025 15:16:35
202103250000        Done    May 02 2025 15:01:41    May 02 2025 16:16:39

******** C48_S2SW_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    May 02 2025 15:01:44    May 02 2025 16:02:12

******** C48_S2SWA_gefs_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    May 02 2025 15:01:46    May 02 2025 16:12:41

******** C96_atm3DVar_extended_3017 ********
202112210000           gfs_fcst_seg1                   194749422             RUNNING                   -         0           0.0
202112210000    gfs_atmos_prod_f339-f360                   194755670             RUNNING                   -         0           0.0
202112210000        gfs_goesupp_f369                   194756453              QUEUED                   -         0           0.0
202112210000        gfs_goesupp_f372                   194756454              QUEUED                   -         0           0.0
202112210000        gfs_goesupp_f375                   194756456              QUEUED                   -         0           0.0
202112210000        gfs_goesupp_f378                   194756457              QUEUED                   -         0           0.0
202112210000    gfs_gempak_f315-f336                   194755675             RUNNING                   -         0           0.0
202112210600           gfs_fcst_seg1                   194753082             RUNNING                   -         0           0.0
202112210600    gfs_atmos_prod_f243-f264                   194756458              QUEUED                   -         0           0.0
202112210600        gfs_goesupp_f261                   194756459              QUEUED                   -         0           0.0
202112210600        gfs_goesupp_f264                   194756461              QUEUED                   -         0           0.0
202112210600        gfs_goesupp_f267                   194756462              QUEUED                   -         0           0.0
202112210600        gfs_goesupp_f270                   194756463              QUEUED                   -         0           0.0
202112210600    gfs_gempak_f147-f168                   194755680             RUNNING                   -         1           0.0
202112210600    gfs_gempak_f195-f216                   194755196             RUNNING                   -         0           0.0
202112210600    gfs_gempak_f219-f240                   194756464              QUEUED                   -         0           0.0
202112211200           gfs_fcst_seg0                   194749770             RUNNING                   -         0           0.0
202112211200    gfs_atmos_prod_f147-f168                   194756466              QUEUED                   -         0           0.0
202112211200        gfs_goesupp_f162                   194756467              QUEUED                   -         0           0.0
202112211200        gfs_goesupp_f165                   194756468              QUEUED                   -         0           0.0
202112211200        gfs_goesupp_f168                   194756469              QUEUED                   -         0           0.0
202112211200    gfs_gempak_f105-f112                   194754780             RUNNING                   -         0           0.0
202112211200    gfs_gempak_f113-f120                   194755202             RUNNING                   -         0           0.0
202112211200    gfs_gempak_f123-f144                   194756470              QUEUED                   -         0           0.0

******** C96C48_hybatmaerosnowDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201200        Done    May 02 2025 15:01:51    May 02 2025 15:21:47
202112201800        Done    May 02 2025 15:01:51    May 02 2025 17:26:48
202112210000        Done    May 02 2025 15:01:51    May 02 2025 17:16:44

******** C96C48_hybatmDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201800        Done    May 02 2025 15:01:53    May 02 2025 15:21:50
202112210000        Done    May 02 2025 15:01:53    May 02 2025 17:26:51
202112210600        Done    May 02 2025 15:01:53    May 02 2025 17:06:47

******** C96C48mx500_S2SW_cyc_gfs_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201200        Done    May 02 2025 15:01:56    May 02 2025 15:21:53
202112201800        Done    May 02 2025 15:01:56    May 02 2025 17:26:55
202112210000        Done    May 02 2025 15:01:56    May 02 2025 18:21:48
202112211800        Done    May 02 2025 15:26:54    May 02 2025 18:21:48

******** C96C48_ufs_hybatmDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202402231800        Done    May 02 2025 15:01:58    May 02 2025 15:21:57
202402240000        Done    May 02 2025 15:01:58    May 02 2025 17:57:09
202402240600        Done    May 02 2025 15:01:58    May 02 2025 17:42:07

******** C96mx100_S2S_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
199405010000        Done    May 02 2025 15:02:00    May 02 2025 16:37:05
```

KateFriedman-NOAA commented:
WCOSS2 CI completed minus a few gempak jobs that hit their walltimes. Still investigating cause but am setting this as passed.
```
Mon May  5 13:04:05 UTC 2025
******** C48_ATM_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    May 02 2025 15:01:37    May 02 2025 15:56:35

******** C48mx500_3DVarAOWCDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103241800        Done    May 02 2025 15:01:39    May 02 2025 15:16:31
202103250000        Done    May 02 2025 15:01:39    May 02 2025 16:36:22

******** C48mx500_hybAOWCDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103241800        Done    May 02 2025 15:01:41    May 02 2025 15:16:35
202103250000        Done    May 02 2025 15:01:41    May 02 2025 16:16:39

******** C48_S2SW_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    May 02 2025 15:01:44    May 02 2025 16:02:12

******** C48_S2SWA_gefs_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    May 02 2025 15:01:46    May 02 2025 16:12:41

******** C96_atm3DVar_extended_3017 ********
202112210600    gfs_gempak_f147-f168                   194755680                DEAD                 -29         2        1853.0
202112211200    gfs_gempak_f123-f144                   194760980                DEAD                 -29         2        1848.0
202112211200    gfs_gempak_f147-f168                   194761997                DEAD                 -29         2        1829.0

******** C96C48_hybatmaerosnowDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201200        Done    May 02 2025 15:01:51    May 02 2025 15:21:47
202112201800        Done    May 02 2025 15:01:51    May 02 2025 17:26:48
202112210000        Done    May 02 2025 15:01:51    May 02 2025 17:16:44

******** C96C48_hybatmDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201800        Done    May 02 2025 15:01:53    May 02 2025 15:21:50
202112210000        Done    May 02 2025 15:01:53    May 02 2025 17:26:51
202112210600        Done    May 02 2025 15:01:53    May 02 2025 17:06:47

******** C96C48mx500_S2SW_cyc_gfs_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201200        Done    May 02 2025 15:01:56    May 02 2025 15:21:53
202112201800        Done    May 02 2025 15:01:56    May 02 2025 17:26:55
202112210000        Done    May 02 2025 15:01:56    May 02 2025 18:21:48
202112211800        Done    May 02 2025 15:26:54    May 02 2025 18:21:48

******** C96C48_ufs_hybatmDA_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202402231800        Done    May 02 2025 15:01:58    May 02 2025 15:21:57
202402240000        Done    May 02 2025 15:01:58    May 02 2025 17:57:09
202402240600        Done    May 02 2025 15:01:58    May 02 2025 17:42:07

******** C96mx100_S2S_3017 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
199405010000        Done    May 02 2025 15:02:00    May 02 2025 16:37:05
```
